### ISSUE TITLE
Merging the emulator to the current scheduler framework service and new RDL

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#48]
Stephen, Suraj and I have just discussed this. To facilitate future scheduling development and research (e.g., dynamic scheduling), we will make it a priority to merge the emulator codes into the latest flux-sched. I am creating this issue to capture the communications needed for this effort.

**Comment by SteVwonder** [comment]
My current plan is to do this in two steps.  Merge the latest emulator code into the master with one PR, then do another PR that makes the scheduler framework service compatible with the emulator and EASY backfilling (through the introduction of extra schedule_jobs functions).

The 1st PR should be pretty straight forward.  It is just taking sometime to squash down 9 months of commits (where each intermediate commit compiles and runs).

**Comment by SteVwonder** [comment]
For the 2nd PR, does it make sense to keep most of the shared emulator code in the `simulator` directory and just have the scheduler framework `#include "simulator.h"`?

Also, I plan to change the `simulator` directory name to `emulator` to correct the misnomer.

**Comment by dongahn** [comment]
The plan makes sense to me. 

Ideally, it would be best if the scheduler framework doesn't know it does its operation for a real environment or an emulated environment. At first glance, it feels the scheduler framework doesn't need to use any functions that belong to the emulator... or if it does the emulator can help faking it. 

Perhaps, we can make it such that one passes "emulation-mode" as an argument to the framework service module, which then uses his flag to decide whether it has to interact with `rexec`. E.g., if this flag is one, we won't launch a program right?

Now, are there emulator specific actions that scheduler framework should do (besides disabling the interaction w/ rexec)?

**Comment by dongahn** [comment]
BTW, if JSC and things need to be adjusted for this (which I suspect we will have to), let's work together to make this happen.

**Comment by SteVwonder** [comment]
I agree on the `emulation-mode` flag, but I believe that the scheduler framework does need to "call out" to shared emulator code.  Particularly to register new events with the emulator manager.

As far as modifications that don't require calling out to shared emulator code:
* As you mentioned, it would also need to communicate with the emulated rexec rather than the real rexec
  * This is easy to fake by unloading rexec and replacing it with the emulated version
* It also needs to queue up resource and job events and handle them when triggered by the emulation manager rather than handling them immediately.
  * Should be easy to do by checking the `emulation-mode` flag and queueing the events in a zlist

What might make the most sense is to clean up my `emulator.h` so that it contains the minimal subset that "emulator-compatible" modules need in order to participate in the emulation.  This would include stuff like:
* `emu_state_t` struct
  * `new_emu_state` function
  * `emu_state_to_json` function
  * `json_to_emu_state` function
  * `free_emu_state` function
* `emu_alive_request` function
* `emu_reply_request` function
* `emu_set_next_event` function

There might be 1 or 2 more things, but ^ should be a pretty close approximation.

**Comment by SteVwonder** [comment]
I am leaning towards this approach because this simple header/library could be used by every module that wanted to be "emulator-compatible".  So the `submit` module, `emu_rexec` module, etc could all link against this.  That way there is one standard way to interface with the emulator manager.  But I also understand your concern about intertwining the scheduler framework with the emulator too much.  Do you have any suggestions for greater separation than this?  I think I'm probably a little too close to the problem to see the bigger picture.

**Comment by dongahn** [comment]
Sorry -- have been helping users to fight data races with my tools. Will get back to you.

**Comment by dongahn** [comment]
@SteVwonder: I think your approach makes sense -- I had to page in a bit on this topic. 

Since the emulation manager would actually on top of scheduler framework service and drive it, it is not entire possible to make the framework service completely agonistic to emulation vs. real environment. 

One thing that we will have to think about is though how we can make JSC play nicely with `emu_rexec` because `emu_rexec` isn't part of flux-core whereas JSC is.  What is the link (or is there any link) between wrexec and emu_rexec? Are you writing to the same lwj directories in KVS as `wrexec` etc. We should discuss a bit to come up with a cohesive plan for this aspect.

The other thing I would like to understand a bit more is how the emulation-submit module is related to the current submit script and submit protocol. This will help me to see if the current submit phase of the scheduler framework will work out of box to support emulation.

**Comment by dongahn** [comment]
Also once flux-core issue #338 is addressed, we want to see if it makes sense to latch onto the new submission protocol for both real scheduling and emulation.

**Comment by dongahn** [comment]
This should also be closed once @SteVwonder's next PR is merged in.

**Comment by dongahn** [comment]
@SteVwonder: I tested with your branch a bit on Cab and found that `t0002-waitjob.t' and `t2000-fcfs` fail with one broker dumping its core. Is your failure mode the same? 

```
lt-flux-broker: src/zmsg.c:466: zmsg_last: Assertion `zmsg_is (self)' failed.
flux-start: 0 (pid 180409) Aborted
flux-start: 1 (pid 180410) Killed
flux-start: 3 (pid 180412) Killed
flux-start: 2 (pid 180411) Killed
cab670{dahn}108: not ok 1 - jsc: expected job-event sequence for single-job scheduling


(gdb) where
#0  0x00002aaaac74f625 in raise (sig=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
#1  0x00002aaaac750e05 in abort () at abort.c:92
#2  0x00002aaaac74874e in __assert_fail_base (fmt=<value optimized out>, assertion=0x2aaaab7e9aea "zmsg_is (self)", file=0x2aaaab7e9adf "src/zmsg.c", line=<value optimized out>, 
    function=<value optimized out>) at assert.c:96
#3  0x00002aaaac748810 in __assert_fail (assertion=0x2aaaab7e9aea "zmsg_is (self)", file=0x2aaaab7e9adf "src/zmsg.c", line=466, function=0x2aaaab7ea759 "zmsg_last") at assert.c:105
#4  0x00002aaaab7b5b7a in zmsg_last () from /usr/lib64/libczmq.so.1
#5  0x00002aaaab11876b in flux_msg_get_type (msg=<value optimized out>, type=0x2aaaed0dd9fc) at message.c:236
#6  0x00002aaaab113b0b in update_tx_stats (h=0x2aaaf0000ab0, msg=0x2aaaf1002b70, flags=0) at handle.c:347
#7  flux_send (h=0x2aaaf0000ab0, msg=0x2aaaf1002b70, flags=0) at handle.c:395
#8  0x00002aaaec22d628 in req_tpexec_exec (h=0x2aaaf0000ab0, job=0x2aaaf158b550) at schedsrv.c:914
#9  0x00002aaaec22d6ff in req_tpexec_run (h=0x2aaaf0000ab0, job=0x2aaaf158b550) at schedsrv.c:943
#10 0x00002aaaec22e78f in action (ctx=0x2aaaf00093b0, job=0x2aaaf158b550, newstate=J_ALLOCATED) at schedsrv.c:1344
#11 0x00002aaaec22eb89 in job_status_cb (jcb=0x2aaaf165dcb0, arg=0x2aaaf0000ab0, errnum=0) at schedsrv.c:1438
#12 0x00002aaaec43fad9 in invoke_cbs (h=0x2aaaf0000ab0, jcb=0x2aaaf165dcb0, j=<value optimized out>, errnum=0) at jstatctl.c:756
#13 0x00002aaaec441697 in job_state_cb (h=0x2aaaf0000ab0, w=<value optimized out>, msg=<value optimized out>, arg=<value optimized out>) at jstatctl.c:828
#14 0x00002aaaab115db5 in handle_cb (r=0x2aaaf00081c0, hw=<value optimized out>, revents=<value optimized out>, arg=0x2aaaf0007de0) at dispatch.c:297
#15 0x00002aaaab12efdd in ev_invoke_pending (loop=0x2aaaf00081e0) at ev.c:3088
#16 0x00002aaaab13360a in ev_run (loop=0x2aaaf00081e0, flags=0) at ev.c:3488
#17 0x00002aaaab115651 in flux_reactor_run (r=0x2aaaf00081c0, flags=0) at reactor.c:122
#18 0x00002aaaec22f030 in mod_main (h=0x2aaaf0000ab0, argc=1, argv=0x2aaaf00081a0) at schedsrv.c:1513
#19 0x000000000040d2a8 in module_thread (arg=0x656790) at module.c:139
#20 0x00002aaaabc739d1 in start_thread (arg=0x2aaaed0de700) at pthread_create.c:301
#21 0x00002aaaac8058fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115
```

**Comment by dongahn** [comment]
@SteVwonder: Ok. Restoring this the original `req_tpexec_exec` function allowed both of these test cases to pass. This leads me to believe the issue is a very simple integration error. I will look further to see if I can create a simple patch so that you can move on.

**Comment by dongahn** [comment]
OK. I've fixed the issues and added some minor cleanup. As I suspected, the issue was caused by a simple integration error. @SteVwonder: I've issue a PR to your fork, please review it and take them in.
